Tests/OpenCombineTests/PublisherTests/FlatMapTests.swift

```diff
@@ -0,0 +1,539 @@
+            $0.request(.max(1))
+        })
+
+        flatMap.subscribe(downstreamSubscriber)
+
+        upstreamPublisher.send(childPublisher)
+
+        // Send a value
+        childPublisher.send(666)
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666)])
+
+        // Buffer a value
+        childPublisher.send(777)
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666)])
+
+        // Fail
+        let error = TestingError.oops
+        upstreamPublisher.send(completion: .failure(error))
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666),
+                                                      .completion(.failure(error))])
+    }
+
+    func testAllSubscriptionsReleasedOnUpstreamFailure() {
+        let upstreamSubscription = CustomSubscription()
+        let upstreamPublisher = CustomPublisherBase<Int, TestingError>(
+            subscription: upstreamSubscription)
+        let downstreamSubscriber = TrackingSubscriberBase<Int, TestingError>(
+            receiveSubscription: { $0.request(.unlimited) })
+
+        let childSubscription = CustomSubscription()
+        let child = CustomPublisher(subscription: childSubscription)
+
+        let flatMap = upstreamPublisher.flatMap { _ in child }
+
+        flatMap.subscribe(downstreamSubscriber)
+
+        XCTAssertEqual(upstreamPublisher.send(1), .none)
+        _ = child.send(666) // TODO: demand
+        upstreamPublisher.send(completion: .failure(TestingError.oops))
+
+        XCTAssertEqual(childSubscription.history, [.requested(.unlimited),
+                                                   .cancelled])
+        XCTAssertEqual(upstreamSubscription.history, [.requested(.unlimited)])
+    }
+
+    func testAllSubscriptionsReleasedOnChildFailure() {
+        let upstreamSubscription = CustomSubscription()
+        let upstreamPublisher = CustomPublisherBase<Int, TestingError>(
+            subscription: upstreamSubscription)
+        let downstreamSubscriber = TrackingSubscriberBase<Int, TestingError>(
+            receiveSubscription: { $0.request(.unlimited) })
+
+        let child1 = PassthroughSubject<Int, TestingError>()
+        let child2Subscription = CustomSubscription()
+        let child2 = CustomPublisher(subscription: child2Subscription)
+
+        let children = [AnyPublisher(child1), AnyPublisher(child2)]
+        let flatMap = upstreamPublisher.flatMap { children[$0] }
+
+        flatMap.subscribe(downstreamSubscriber)
+
+        XCTAssertEqual(upstreamPublisher.send(0), .none)
+        XCTAssertEqual(upstreamPublisher.send(1), .none)
+        _ = child1.send(666)    // TODO: demand
+        _ = child2.send(777)    // TODO: demand
+        child1.send(completion: .failure(TestingError.oops))
```

Comment:
Doesn't it violate the contract? AFAIK, the upstream should wait for all the current children to finish before sending the next value. Also, `child2` should not send any values before `child1` is finished.

If it is intentional, can you add a comment?